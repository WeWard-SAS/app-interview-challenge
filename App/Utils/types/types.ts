import { Moment } from 'moment';
import { StepDataPoint } from '../services/healthService';

export interface ControlsProps {
  month: string;
  updateDate: (date: Moment) => void;
  isLoading: boolean;
}

export type WeekData = {
  key: number;
  days: DayStep[];
};

export interface StepData {
  steps: StepDataPoint[];
  months: string;
  weeks: WeekData[];
}

export interface StepHistory {
  [date: string]: {
    startDate: string;
    endDate: string;
    value: number;
  };
}

export interface DayStep {
  dayNumber: string | number;
  key: number;
  isCurrentMonth: boolean;
  numberOfSteps?: string;
  isDayAfter: boolean;
}
