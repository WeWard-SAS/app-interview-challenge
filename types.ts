export type StepsData = Map<number, number>;
export type StepDataPoint = {
  startDate: string; // une string ISO 8601
  endDate: string; // une string ISO 8601
  value: number;
};
