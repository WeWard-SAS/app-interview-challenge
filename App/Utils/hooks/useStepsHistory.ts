import { useEffect, useState } from 'react';
import { getStepsHistory, isDayAfter } from '../services/healthService';
import moment from 'moment';
import { NativeModules, Platform } from 'react-native';
import { DayStep, StepData, StepHistory, WeekData } from '../types/types';

const isAndroid = Platform.OS === 'android';
const locales = isAndroid
  ? NativeModules.I18nManager.localeIdentifier
  : NativeModules.SettingsManager.settings.AppleLocale ||
    NativeModules.SettingsManager.settings.AppleLanguages[0];

// Current Month will have this format 'YYYY MM'
const useStepsHistory = (currentMonth: string) => {
  const [loading, setLoading] = useState(false);
  const [data, setData] = useState<StepData>();
  const [error, setError] = useState('');

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      try {
        const steps = await getStepsHistory(
          moment(currentMonth, 'YYYY MM').format('YYYY-MM-DD'),
          moment(currentMonth, 'YYYY MM').add(1, 'month').format('YYYY-MM-DD'),
        );
        const result = steps.reduce((acc: StepHistory, curr) => {
          // Formatting the date
          const date = moment(curr.startDate).format('YYYY-MM-DD');

          // Checking if others values exisiting
          if (!acc[date]) {
            // Creating a new entries if not
            acc[date] = {
              startDate: curr.startDate,
              endDate: curr.endDate,
              value: curr.value,
            };
          } else {
            // Updating value (steps) if yes
            acc[date].value += curr.value;
          }

          return acc;
        }, {});

        const groupedStepsData = Object.values(result);
        const daysInMonth = moment(currentMonth, 'YYYY MM').daysInMonth();
        const firstDayOfMonth = moment(currentMonth, 'YYYY MM')
          .startOf('month')
          .day();

        const weeksInMonth = Math.ceil((daysInMonth + firstDayOfMonth) / 7);
        const weeks: WeekData[] = [];

        // Mapping the weeks
        for (let week = 0; week < weeksInMonth; week++) {
          const days: DayStep[] = [];

          // Mapping the days
          for (let day = 0; day < 7; day++) {
            const dayIndex = week * 7 + day - firstDayOfMonth + 1;

            // If day part of the month, pushing steps and index
            if (dayIndex > 0 && dayIndex <= daysInMonth) {
              days.push({
                key: dayIndex,
                dayNumber: dayIndex,
                isCurrentMonth: true,
                numberOfSteps: isDayAfter(
                  moment(`${currentMonth} ${dayIndex}`, 'YYYY MM D'),
                )
                  ? '300'
                  : groupedStepsData
                      .find(item => moment(item.startDate).date() === dayIndex)
                      ?.value.toLocaleString(locales.split('_')[0], {
                        maximumFractionDigits: 0,
                        minimumFractionDigits: 0,
                        useGrouping: true,
                      }),
                isDayAfter: isDayAfter(
                  moment(`${currentMonth} ${dayIndex}`, 'YYYY MM D'),
                ),
              });
            } else {
              days.push({
                key: dayIndex,
                dayNumber: '',
                numberOfSteps: '',
                isCurrentMonth: false,
                isDayAfter: isDayAfter(
                  moment(`${currentMonth} ${dayIndex}`, 'YYYY MM D'),
                ),
              });
            }
          }

          // Updating array with steps values
          weeks.push({
            key: week,
            days: days,
          });
        }

        setData({ steps, months: currentMonth, weeks });
      } catch (err: unknown) {
        if (err instanceof Error) {
          setError(err.message);
        }
      }
      setLoading(false);
    };
    fetchData();
  }, [currentMonth]);

  return { data, error, loading };
};

export default useStepsHistory;
