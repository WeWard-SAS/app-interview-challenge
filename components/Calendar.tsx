import React from 'react';
import {View, Text, StyleSheet, Image, TouchableOpacity} from 'react-native';
import moment from 'moment';
import {EMPTY_STRING} from '../Constants';

type CalendarProps = {
  startDate: Date;
  onStartDateChanged: (newStartDate: Date) => void;
};

const Calendar = (props: CalendarProps) => {
  const {startDate, onStartDateChanged} = props;

  const Header = () => {
    let title = EMPTY_STRING;
    if (startDate) {
      const monthName = moment(startDate).format('MMMM');
      const year = moment(startDate).year();
      title = `${monthName} ${year}`;
    }

    const handleMonthChange = (value: number) => {
      const newStartDate = moment(startDate).add(value, 'month');
      onStartDateChanged(newStartDate.toDate());
    };

    return (
      <View style={styles.headerContainer}>
        <View style={styles.headerTextContainer}>
          <Text style={styles.headerTitle}>{title}</Text>
        </View>
        <View style={styles.navigationContainer}>
          <TouchableOpacity
            style={styles.navigationButton}
            onPress={() => {
              handleMonthChange(-1);
            }}>
            <Image
              source={require('../assets/previous.png')}
              style={styles.navigationImage}
            />
          </TouchableOpacity>
          <TouchableOpacity
            style={styles.navigationButton}
            onPress={() => {
              handleMonthChange(+1);
            }}>
            <Image
              source={require('../assets/next.png')}
              style={styles.navigationImage}
            />
          </TouchableOpacity>
        </View>
      </View>
    );
  };

  const CalendarBody = () => {
    return (
      <View style={styles.calendarContainer}>
        <Text>Calendar Body</Text>
      </View>
    );
  };

  return (
    <View style={styles.mainContainer}>
      <Header />
      <CalendarBody />
    </View>
  );
};

const styles = StyleSheet.create({
  mainContainer: {
    flex: 1,
  },
  headerContainer: {
    flex: 1,
    flexDirection: 'row',
    marginBottom: 30,
  },
  calendarContainer: {
    backgroundColor: 'cyan',
  },
  headerTextContainer: {
    width: '75%',
  },
  headerTitle: {
    fontSize: 20,
    color: 'black',
  },
  navigationContainer: {
    width: '25%',
    flexDirection: 'row',
    alignContent: 'space-between',
  },
  navigationButton: {
    borderWidth: 0,
    marginRight: 10,
    marginLeft: 10,
  },
  navigationImage: {
    width: 25,
    height: 25,
  },
});
export default Calendar;
