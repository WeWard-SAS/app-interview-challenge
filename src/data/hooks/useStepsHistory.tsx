import moment from 'moment';
import {useEffect, useState} from 'react';
import CalendarHelper from '../../helpers/calendarHelper';
import {StepDataPoint, getStepsHistory} from '../services/healthService';

export type StepsHistory = {[key: string]: DayStepDataPoint[]};

export interface DayStepDataPoint {
  date: string;
  value: number;
}

interface MonthData {
  date: string;
  data: DayStepDataPoint[];
}

type UseStepsHistoryHook = [
  moment.Moment,
  StepsHistory,
  boolean,
  () => void,
  () => void,
];

/**
 *
 * @param offset - Determines how many months around the current month should be loaded. Default is 0, so only the current month will be loaded.
 * @returns {UseStepsHistoryHook}
 */
const useStepsHistory = (offset: number = 0): UseStepsHistoryHook => {
  if (offset < 0) {
    throw 'offset cannot be < 0';
  } else if (offset > 10) {
    throw 'offset cannot be > 10';
  }

  const [month, setMonth] = useState(moment(24, 'HH').set('date', 1));

  const [stepsHistory, setStepsHistory] = useState<StepsHistory>({});
  const [isLoading, setIsLoading] = useState(false);

  const selectPreviousMonth = (): void =>
    setMonth(moment(month).subtract(1, 'month'));

  const selectNextMonth = (): void => setMonth(moment(month).add(1, 'month'));

  useEffect(() => {
    (async function () {
      const alreadyLoadedMonths = Object.keys(stepsHistory);
      const monthsOffsetArray = [...Array(offset).keys()];

      // Determine number of months to load.
      // If offset is '4', load 4 months before and after current month
      // and current month
      let monthsToLoad = [
        ...monthsOffsetArray
          .map(i => [
            moment(month)
              .subtract(i + 1, 'month')
              .format(CalendarHelper.DATE_FORMAT),
            moment(month)
              .add(i + 1, 'month')
              .format(CalendarHelper.DATE_FORMAT),
          ])
          .flat(),
        month.format(CalendarHelper.DATE_FORMAT),
      ];

      monthsToLoad = monthsToLoad.filter(
        (month: string) => !alreadyLoadedMonths.includes(month),
      );

      if (monthsToLoad.length) {
        setIsLoading(true);
      }

      const getMonthData = async (month: moment.Moment): Promise<MonthData> => {
        const startDate = (): moment.Moment =>
          moment(month).subtract(0, 'month');

        const endDate = (): moment.Moment => moment(month).add(1, 'month');

        const steps: StepDataPoint[] = await getStepsHistory(
          startDate().toISOString(),
          endDate().toISOString(),
        );

        let dayData: DayStepDataPoint[] = [];

        steps.forEach((step: StepDataPoint) => {
          let key: string = moment(step.startDate).format(
            CalendarHelper.DATE_FORMAT,
          );

          let data: DayStepDataPoint | undefined = dayData.find(
            day => day.date === key,
          );

          if (!data) {
            data = {
              date: key,
              value: 0,
            };
          }

          data.value = data.value + step.value;
          dayData.push(data);
        });

        const monthData: MonthData = {
          date: month.format(CalendarHelper.DATE_FORMAT),
          data: dayData,
        };

        return monthData;
      };

      const monthsData: MonthData[] = await Promise.all(
        monthsToLoad.map((month: string) => getMonthData(moment(month))),
      );

      const newHistory: StepsHistory = Object.fromEntries(
        monthsData.map(({date, data}) => [date, data]),
      );

      setStepsHistory(Object.assign({}, newHistory, stepsHistory));

      if (monthsToLoad.length) {
        setIsLoading(false);
      }
    })();
  }, [month]);

  return [month, stepsHistory, isLoading, selectPreviousMonth, selectNextMonth];
};

export default useStepsHistory;
