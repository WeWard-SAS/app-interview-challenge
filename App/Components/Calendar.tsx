import React, { useState } from 'react';
import { ActivityIndicator, Text, View } from 'react-native';
import moment, { Moment } from 'moment';

import useStepsHistory from '../Utils/hooks/useStepsHistory';
import ApplicationStyles from '../styles/ApplicationStyle';
import Metrics from '../styles/Metrics';

import CalendarControls from './CalendarControls';
import CalendarWeekDays from './CalendarWeekDays';

const Calendar = () => {
  const [currentDate, setDate] = useState<string>(
    moment().startOf('month').format('YYYY MM'),
  );

  const { data, loading, error } = useStepsHistory(currentDate);

  if (error.length > 0) {
    return (
      <View style={[ApplicationStyles.flex, ApplicationStyles.columnContainer]}>
        <Text
          style={[ApplicationStyles.genericText, ApplicationStyles.textError]}>
          {error}
        </Text>
      </View>
    );
  } else if (data) {
    return (
      <View style={ApplicationStyles.flex}>
        <CalendarControls
          month={data.months}
          updateDate={(date: Moment) => setDate(date.format('YYYY MM'))}
          isLoading={loading}
        />
        <View style={ApplicationStyles.divider} />
        {loading ? (
          <ActivityIndicator
            size="small"
            style={{ marginTop: Metrics.doublePadding }}
          />
        ) : (
          <CalendarWeekDays weeks={data.weeks} />
        )}
      </View>
    );
  }
  return <></>;
};

export default Calendar;
