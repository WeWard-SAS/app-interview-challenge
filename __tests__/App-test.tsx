import React from 'react';
import renderer, {
  ReactTestInstance,
  ReactTestRenderer,
} from 'react-test-renderer';

import useStepsHistory from '../App/Utils/hooks/useStepsHistory';
import { StepData } from '../App/Utils/types/types';

import Calendar from '../App/Components/Calendar';

jest.mock('../App/Utils/hooks/useStepsHistory');

describe('Calendar', () => {
  const mockedUseStepsHistory = useStepsHistory as jest.MockedFunction<
    typeof useStepsHistory
  >;
  const data: StepData = {
    months: '2022 04',
    weeks: [],
  };
  test('renders correctly an error message if hook returned one', async () => {
    mockedUseStepsHistory.mockReturnValue({
      data,
      loading: false,
      error: 'An error occured',
    });

    const tree: ReactTestRenderer = renderer.create(<Calendar />);
    const instance: ReactTestInstance = tree.root;
    expect(
      instance.findByProps({ testID: 'error-message' }).children,
    ).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });
  test('renders correctly with loader, data being fetched', async () => {
    mockedUseStepsHistory.mockReturnValue({
      data,
      loading: true,
      error: '',
    });

    const tree: ReactTestRenderer = renderer.create(<Calendar />);
    const instance: ReactTestInstance = tree.root;
    expect(() =>
      instance.findByProps({ testID: 'error-message' }),
    ).toThrowError('No instances found with props: {"testID":"error-message"}');
    expect(() => instance.findByProps({ testID: 'loader' })).toBeTruthy();
    expect(
      instance.findByProps({ testID: 'calendar-controls' }).children,
    ).toBeTruthy();

    expect(tree).toMatchSnapshot();
  });
  test('renders correctly with data having been fetched and no loader displayed', async () => {
    mockedUseStepsHistory.mockReturnValue({
      data: {
        months: '2023 03',
        weeks: [],
      },
      loading: false,
      error: '',
    });

    const tree: ReactTestRenderer = renderer.create(<Calendar />);
    const instance: ReactTestInstance = tree.root;

    expect(() =>
      instance.findByProps({ testID: 'error-message' }),
    ).toThrowError('No instances found with props: {"testID":"error-message"}');
    expect(
      instance.findByProps({ testID: 'calendar-controls' }).children,
    ).toBeTruthy();
    expect(
      instance.findByProps({ testID: 'calendar-controls' }).props.month,
    ).toEqual('2023 03');
    expect(
      instance.findByProps({ testID: 'week-days-data' }).children,
    ).toBeTruthy();

    expect(tree).toMatchSnapshot();
  });
});
