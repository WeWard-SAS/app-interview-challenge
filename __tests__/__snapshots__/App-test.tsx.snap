// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders correctly an error message if hook returned one 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "OpenSans-SemiBold",
        },
        Object {
          "color": "#f0834b",
          "fontSize": 12,
        },
      ]
    }
    testID="error-message"
  >
    An error occured
  </Text>
</View>
`;

exports[`Calendar renders correctly with data having been fetched and no loader displayed 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
    testID="calendar-controls"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "OpenSans-SemiBold",
          },
          Object {
            "color": "#12192b",
            "fontSize": 17,
          },
        ]
      }
    >
      Mars 2023
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginRight": 24,
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../App/assets/left-chevron.png",
            }
          }
          style={
            Object {
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../App/assets/right-chevron.png",
            }
          }
          style={
            Object {
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#a9b2c3",
        "height": 2,
        "marginTop": 24,
        "opacity": 0.1,
      }
    }
  />
  <View
    testID="week-days-data"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "marginTop": 12,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          LUN
          .
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          MAR
          .
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          MER
          .
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          JEU
          .
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          VEN
          .
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          SAM
          .
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          DIM
          .
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Calendar renders correctly with loader, data being fetched 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
    testID="calendar-controls"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "OpenSans-SemiBold",
          },
          Object {
            "color": "#12192b",
            "fontSize": 17,
          },
        ]
      }
    >
      Avril 2022
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginRight": 24,
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../App/assets/left-chevron.png",
            }
          }
          style={
            Object {
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../App/assets/right-chevron.png",
            }
          }
          style={
            Object {
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#a9b2c3",
        "height": 2,
        "marginTop": 24,
        "opacity": 0.1,
      }
    }
  />
  <ActivityIndicator
    size="small"
    style={
      Object {
        "marginTop": 24,
      }
    }
    testID="loader"
  />
</View>
`;
